### DISCUSSION TITLE
[FR]: Job ordering

### DISCUSSION THREAD

**Discussion by ewhauser** [riverqueue/river#431] [Ideas]
In some systems, you need to guarantee order across a given partition key while still having multiple queue consumer threads. SQS has the concept of [FIFO queues](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-fifo-queues.html) that we currently use for this.

The canonical use case a high volume SMS system where you are sending multiple messages to a specific customer and always want to send them in order.

**Comment by brandur** [comment]
This does seem like something that might be useful in the project — I've seen the same kind of thing elsewhere too.

I'm not sure we're likely to get to this right away, so I'd suggest this workaround in the meantime: using the SMS example again, instead of putting the SMS messages directly in a River job, hold them instead as a separate database model that's ordered naturally on insert order. Then, your River job runs and just clears whatever SMS messages it finds in the database:

``` go
func (w *SendSMSWorker) Work(ctx context.Context, job *river.Job[SendSMSArgs]) error {
    ...

    messagesToSend, err := queries.SMSMessageLoadByUser(job.UserID)
    ...

    for _, message := range messagesToSend {
        // send SMS message
    }

    return nil
}
```

**Reply by fsaintjacques** [reply]
The issue is with concurrent workers draining from the same partition. You can hack this with postgres' advisory locks but now you've entered deadlock territory or simply highly unbalanced partitions (hello multi-tenant) can bring the system to a halt. 

The goal here is that workers handle this transparently for you.

**Comment by bgentry** [comment]
FYI, this functionality is now part of [sequences](https://riverqueue.com/docs/pro/sequences) in River Pro ✨

**Reply by sharlenaluyen** [reply]
Is there code or examples regarding sequences in the River repo that you can link me to, per chance?

**Reply by bgentry** [reply]
The best source is the docs: https://riverqueue.com/docs/pro/sequences

The riverpro Go package docs don't currently have an executable example for sequences.
